service/kernel/data_management/train_test_split.cpp

```diff
@@ -0,0 +1,203 @@
+    const size_t nTrainBlocks = nTrainRows / BLOCK_SIZE ? nTrainRows / BLOCK_SIZE : 1;
+    const size_t nTestBlocks = nTestRows / BLOCK_SIZE ? nTestRows / BLOCK_SIZE : 1;
+    const int nTrainSurplus = nTrainRows / BLOCK_SIZE ? nTrainRows % BLOCK_SIZE : nTrainRows - BLOCK_SIZE;
+    const int nTestSurplus = nTestRows / BLOCK_SIZE ? nTestRows % BLOCK_SIZE : nTestRows - BLOCK_SIZE;
+
+    BlockDescriptor<DataType> origBlock, trainBlock, testBlock;
+    inputTable.getBlockOfRows(0, nTrainRows + nTestRows, readOnly, origBlock);
+    trainTable.getBlockOfRows(0, nTrainRows, readWrite, trainBlock);
+    testTable.getBlockOfRows(0, nTestRows, readWrite, testBlock);
+    const DataType * origDataPtr = origBlock.getBlockPtr();
+    DataType * trainDataPtr = trainBlock.getBlockPtr();
+    DataType * testDataPtr = testBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(origDataPtr);
+    DAAL_CHECK_MALLOC(trainDataPtr);
+    DAAL_CHECK_MALLOC(testDataPtr);
+
+    runBlocks(nTrainRows * nColumns > THREADING_BORDER, nTrainBlocks, [&](size_t iBlock) {
+        const size_t start = iBlock * BLOCK_SIZE;
+        const size_t end   = iBlock == nTrainBlocks - 1 ? start + BLOCK_SIZE + nTrainSurplus : start + BLOCK_SIZE;
+
+        for (size_t i = start; i < end; ++i)
+            for (size_t j = 0; j < nColumns; ++j)
+                trainDataPtr[i * nColumns + j] = origDataPtr[trainIdx[i] * nColumns + j];
+    });
+
+    runBlocks(nTestRows * nColumns > THREADING_BORDER, nTestBlocks, [&](size_t iBlock) {
+        const size_t start = iBlock * BLOCK_SIZE;
+        const size_t end   = iBlock == nTestBlocks - 1 ? start + BLOCK_SIZE + nTestSurplus : start + BLOCK_SIZE;
+
+        for (size_t i = start; i < end; ++i)
+            for (size_t j = 0; j < nColumns; ++j)
+                testDataPtr[i * nColumns + j] = origDataPtr[testIdx[i] * nColumns + j];
+    });
+    return s;
+}
+
+
+template <typename IdxType>
+services::Status trainTestSplitImpl(NumericTable & inputTable, NumericTable & trainTable, NumericTable & testTable,
+                                    NumericTable & trainIdxTable, NumericTable & testIdxTable, NumericTable & columnTypesTable)
+{
+    services::Status s;
+    const NTLayout layout  = inputTable.getDataLayout();
+    const size_t nColumns = trainTable.getNumberOfColumns();
+    const size_t nTrainRows = trainTable.getNumberOfRows();
+    const size_t nTestRows = testTable.getNumberOfRows();
+    DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, nTrainRows + nTestRows, nColumns);
+
+    BlockDescriptor<int> columnTypesBlock;
+    columnTypesTable.getBlockOfRows(0, nColumns, readOnly, columnTypesBlock);
+    const int * columnTypesPtr = columnTypesBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(columnTypesPtr);
+
+    BlockDescriptor<IdxType> trainIdxBlock, testIdxBlock;
+    trainIdxTable.getBlockOfColumnValues(0, 0, nTrainRows, readOnly, trainIdxBlock);
+    testIdxTable.getBlockOfColumnValues(0, 0, nTestRows, readOnly, testIdxBlock);
```

Comment:
Better to get Values in parallel to avoid large conversion overheads

Reply:
Parallel getting is added where it's possible